In [54]:
import pandas as pd
import numpy as np

In [10]:
df = pd.read_csv('https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv')

In [11]:
df.head()

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                  170            3.0       159.0     3413.433759   
1                  130            5.0        97.0     3149.664934   
2                  170            NaN        78.0     3079.038997   
3                  220            4.0         NaN     2542.392402   
4                  210            1.0       140.0     3460.870990   

   acceleration  model_year  origin fuel_type         drivetrain  num_doors  \
0          17.7        2003  Europe  Gasoline    All-wheel drive        0.0   
1          17.8        2007     USA  Gasoline  Front-wheel drive        0.0   
2          15.1        2018  Europe  Gasoline  Front-wheel drive        0.0   
3          20.2        2009     USA    Diesel    All-wheel drive        2.0   
4          14.4        2009  Europe  Gasoline    All-wheel drive        2.0   

   fuel_efficiency_mpg  
0            13.231729  
1            13.688217  
2            14.246341  
3            16.912736  
4            12.488369

# Q1
### What version of Pandas did you install?

In [12]:
pd.__version__

'2.3.1'

# Q2
### How many records are in the dataset?

In [13]:
df.shape

(9704, 11)

# Q3
### How many fuel types are presented in the dataset?

In [14]:
df['fuel_type'].nunique()

2

# Q4
### How many columns in the dataset have missing values?

In [15]:
df.isna().sum()
# 4

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

# Q5 
### What's the maximum fuel efficiency of cars from Asia?

In [22]:
df[df['origin']=='Asia']['fuel_efficiency_mpg'].max()

np.float64(23.759122836520497)

# Q6

Find the median value of the horsepower column in the dataset. <br>
Next, calculate the most frequent value of the same horsepower column.<br>
Use the fillna method to fill the missing values in the horsepower column with the most frequent value from the previous step.<br>
Now, calculate the median value of horsepower once again.<br>


In [23]:
df['horsepower'].median()

np.float64(149.0)

In [32]:
df['horsepower'].mode()[0]

np.float64(152.0)

In [33]:
df['horsepower'] = df['horsepower'].fillna(df['horsepower'].mode()[0])

In [35]:
df['horsepower'].median()

np.float64(152.0)

The median horsepower increase from 149 to 152

# Q7

Select all the cars from Asia <br>
Select only columns vehicle_weight and model_year <br>
Select the first 7 values <br>
Get the underlying NumPy array. Let's call it X. <br>
Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.<br>
Invert XTX. <br>
Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].<br>
Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.<br>
What's the sum of all the elements of the result?<br>

In [43]:
df_asia = df[df['origin']=='Asia'][['vehicle_weight','model_year']].reset_index(drop=True)[0:7]

In [63]:
X = df_asia.to_numpy()

In [64]:
X

array([[2714.21930965, 2016.        ],
       [2783.86897424, 2010.        ],
       [3582.68736772, 2007.        ],
       [2231.8081416 , 2011.        ],
       [2659.43145076, 2016.        ],
       [2844.22753389, 2014.        ],
       [3761.99403819, 2019.        ]])

In [65]:
XTX = X.T @ X

In [66]:
XTX

array([[62248334.33150762, 41431216.5073268 ],
       [41431216.5073268 , 28373339.        ]])

In [67]:
XTX_inv = np.linalg.inv(XTX)

In [68]:
XTX_inv

array([[ 5.71497081e-07, -8.34509443e-07],
       [-8.34509443e-07,  1.25380877e-06]])

In [69]:
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])

In [70]:
y

array([1100, 1300,  800,  900, 1000, 1100, 1200])

In [71]:
w = (XTX_inv @ X.T) @ y

In [72]:
w

array([0.01386421, 0.5049067 ])

In [73]:
w.sum()

np.float64(0.5187709081074016)